### PR TITLE
feat: Added support for account level governance of AI Monitoring

### LINKED ISSUE #2325: Add support for account disablement of AI Monitoring
There is a feature request to add support for disabling AI Monitoring at the account level. This request has been given elevated priority across several agents, with the Node.js agent being one of them.

We need to:

1. Update our connect service interaction code to recognize the new setting being sent by the connect service
2. Apply the setting to the current agent instance
3. Verify the setting is working via the staging environment

The updated spec for the payload being sent by the connect service can be viewed in our GHE instance. In short, to disable AI Monitoring the connect service will send:

```json
{
	"return_value": {
		"collect_ai": false
	}
}
```

### PR DESCRIPTION
**Pull request by jsumners-nr**
This PR resolves #2325.

### CODE REVIEW

File: lib/instrumentation/openai.js [hunk 1]
```diff
@@ -99,12 +99,18 @@ function addLlmMeta({ agent, segment }) {
...
  * @param {object} params input params
  * @param {Agent} params.agent NR agent instance
+ * @param {Shim} params.shim the current shim instance
  * @param {TraceSegment} params.segment active segment from chat completion
  * @param {object} params.request chat completion params
  * @param {object} params.response chat completion response
  * @param {boolean} [params.err] err if it exists
  */
-function recordChatCompletionMessages({ agent, segment, request, response, err }) {
+function recordChatCompletionMessages({ agent, shim, segment, request, response, err }) {
```
Comment (bizob2828) [inline lib/instrumentation/openai.js:108]: this file also has instrumentation to record embeddings. Right now it's still doing that

File: lib/instrumentation/langchain/common.js [hunk 2]
```diff
@@ -43,11 +43,17 @@ common.mergeMetadata = function mergeMetadata(localMeta = {}, paramsMeta = {}) {
...
  * @param {object} params function params
  * @param {Agent} params.agent NR agent
+ * @param {Shim} params.shim current shim instance
  * @param {string} params.type type of llm event(i.e.- LlmChatCompletionMessage, LlmTool, etc)
  * @param {object} params.msg the llm event getting enqueued
  * @param {string} params.pkgVersion version of langchain library instrumented
  */
-common.recordEvent = function recordEvent({ agent, type, msg, pkgVersion }) {
+common.recordEvent = function recordEvent({ agent, shim, type, msg, pkgVersion }) {
+  if (common.shouldSkipInstrumentation(agent.config) === true) {
```
Comment (bizob2828) [inline lib/instrumentation/langchain/common.js:52]: this behavior is different than the other libraries so when ai_monitoring.enabled is false it still creates the events in memory but doesn't do anything with them.

Comment (bizob2828) [reply lib/instrumentation/langchain/common.js:52]: there are cases where it is creating LLM events like `LlmErrorMessage`. It's also doing a lot of parsing. I would change langchain instrumentation and check for this flag before it attempts to event create any llm events

File: test/unit/config/config-server-side.test.js [hunk 3]
```diff
@@ -161,6 +161,15 @@ tap.test('when receiving server-side configuration', (t) => {
     t.end()
   })
 
+  t.test('should disable ai monitoring', (t) => {
+    // Default is `false`, but let's verify it anyway.
+    t.equal(config.ai_monitoring.enabled, false)
```
Comment (bizob2828) [inline test/unit/config/config-server-side.test.js:166]: shouldn't this be set to `true`?

**Review by bizob2828 — CHANGES_REQUESTED** [review]
I tested with openai. I had a few comments about langchain and some tests

File: test/unit/config/config-server-side.test.js (hunk 3 above)
Comment (jsumners-nr) [reply test/unit/config/config-server-side.test.js:166]: The default is `false` https://github.com/newrelic/node-newrelic/blob/7cd6aef7ecc32d8647f0313b36be368bfa14a0e9/lib/config/default.js#L1360

Comment (bizob2828) [reply test/unit/config/config-server-side.test.js:166]: yea i know but you're basically setting it to false and then getting false back from connect and asserting ai_monitoring.enabled is still false, I don't see what the test is actually asserting

File: lib/instrumentation/openai.js [hunk 4]
```diff
@@ -289,6 +298,11 @@ module.exports = function initialize(agent, openai, moduleName, shim) {
     openai.Embeddings.prototype,
     'create',
     function wrapEmbeddingCreate(shim, embeddingCreate, name, args) {
+      if (shouldSkipInstrumentation(agent.config, shim) === true) {
```
Comment (bizob2828) [inline lib/instrumentation/openai.js:301]: same comment as in langchain. this is happening too early and it won't properly record the span

```
--- a/lib/instrumentation/openai.js
+++ b/lib/instrumentation/openai.js
@@ -298,11 +298,6 @@ module.exports = function initialize(agent, openai, moduleName, shim) {
     openai.Embeddings.prototype,
     'create',
     function wrapEmbeddingCreate(shim, embeddingCreate, name, args) {
-      if (shouldSkipInstrumentation(agent.config, shim) === true) {
-        shim.logger.debug('skipping sending of ai data')
-        return
-      }
-
       const [request] = args
       return new RecorderSpec({
         name: OPENAI.EMBEDDING,
@@ -315,10 +310,16 @@ module.exports = function initialize(agent, openai, moduleName, shim) {
             // In that case, we define it to be an empty object.
             response = {}
           }
+          segment.end()
+
+          if (shouldSkipInstrumentation(agent.config, shim) === true) {
+            shim.logger.debug('skipping sending of ai data')
+            return
+          }
+
           response.headers = segment[openAiHeaders]
           // explicitly end segment to get consistent duration
           // for both LLM events and the segment
-          segment.end()
           const embedding = new LlmEmbedding({
             agent,
             segment,
```

take note i moved ending segment before assigning the segment to a symbol on headers

**Review by bizob2828 — CHANGES_REQUESTED** [review]
comments on location of config check to still create spans and end them but refrain from creating LLM events

**Review by bizob2828 — APPROVED** [review]

**Comment by jsumners-nr** [comment]
I'm going to wait to merge this until the team is ready to ship the feature

**Pull request merged**